runtime/wasm/src/module/mod.rs

```diff
@@ -269,126 +123,367 @@ impl WasmiModule {
+            // This task is likely to outlive the instance, which is fine.
+            let interrupt_handle = instance.store().interrupt_handle().unwrap();
+            let should_interrupt = should_interrupt.clone();
+            graph::spawn(async move {
+                tokio::time::delay_for(timeout).await;
+                if should_interrupt.load(Ordering::SeqCst) {
+                    interrupt_handle.interrupt()
+                }
+            });
+        }
+
+        *shared_instance.borrow_mut() = Some(WasmInstance {
+            instance,
+            memory_allocate: Box::new(memory_allocate),
+            memory,
+            ctx,
+            valid_module,
+            host_metrics,
+            timeout,
+            should_interrupt,
+
+            // `arena_start_ptr` will be set on the first call to `raw_new`.
+            arena_free_size: 0,
+            arena_start_ptr: 0,
+        });
 
-        result.map(|_| self.ctx.state).map_err(|err| {
-            format_err!(
-                "Failed to handle Ethereum block with handler \"{}\": {}",
-                handler_name,
-                format_wasmi_error(err)
-            )
+        Ok(WasmInstanceHandle {
+            instance: shared_instance.clone(),
         })
     }
 }
 
-impl AscHeap for WasmiModule {
-    fn raw_new(&mut self, bytes: &[u8]) -> Result<u32, Error> {
+impl AscHeap for WasmInstance {
+    fn raw_new(&mut self, bytes: &[u8]) -> u32 {
         // We request large chunks from the AssemblyScript allocator to use as arenas that we
         // manage directly.
 
-        static MIN_ARENA_SIZE: u32 = 10_000;
+        static MIN_ARENA_SIZE: i32 = 10_000;
 
-        let size = u32::try_from(bytes.len()).unwrap();
+        let size = i32::try_from(bytes.len()).unwrap();
         if size > self.arena_free_size {
             // Allocate a new arena. Any free space left in the previous arena is left unused. This
             // causes at most half of memory to be wasted, which is acceptable.
             let arena_size = size.max(MIN_ARENA_SIZE);
-            let allocated_ptr = self
-                .module
-                .clone()
-                .invoke_export("memory.allocate", &[RuntimeValue::from(arena_size)], self)
-                .expect("Failed to invoke memory allocation function")
-                .expect("Function did not return a value")
-                .try_into::<u32>()
-                .expect("Function did not return u32");
-            self.arena_start_ptr = allocated_ptr;
+            self.arena_start_ptr = (self.memory_allocate)(arena_size).unwrap();
             self.arena_free_size = arena_size;
         };
 
-        let ptr = self.arena_start_ptr;
-        self.memory.set(ptr, bytes)?;
+        let ptr = self.arena_start_ptr as usize;
+
+        // Safe because we are accessing and immediately dropping the reference to the data.
+        unsafe { self.memory.data_unchecked_mut()[ptr..(ptr + bytes.len())].copy_from_slice(bytes) }
```

Comment:
I audited this for safety and I think it checks out. The situation may be a bit more tenuous than the comment implies though. The comment implies that the safety of this operation can be audited locally as if only the code within the `unsafe` block matters. This is not true.

Unfortunately, all the code which has access to the path `WasmInstance::memory: Memory` (which includes the set of all code which might access `WasmInstance::instance: wasmtime::Instance` and `WasmInstance::memory_allocate`) is a part of the surface area of the unsafety because these are paths that an additional reference (shared or not) could be created which aliases this borrow. Note that the `data_unchecked_mut` method doesn't even require `&mut self`! This is one of the most unsafe API's I've seen in the wild in some time.

This is concerning because these data are accessible by host exports! This is critical to security because this is exactly the kind of place for an attack vector to try to read memory owned by the indexer (like a private key, which we need to make sure an indexer does not have).

Is there a way to sandbox all paths to this memory in a way that less code has to be audited? Ideally, we could drop the module and the memory into something rock-solid with less than 100 lines of surface area.

Reply:
This comment really is insufficient, so I'll explain my reasoning. First instance and memory is `!Sync`, at least for now, so these unsafe blocks can't be called concurrently.

Given that, the only way this API can cause unsoundness outside an `unsafe` block is if that block leaks a reference to the memory to safe code. We can adopt as an invariant that no unsafe block does this, documenting and auditing in the two unsafe blocks we have. The beauty being that if this invariant is locally true for all unsafe blocks, then all our code is safe.

Auditing all the paths that use `Instance`, tracking usage of memory for these paths and concluding that the code really is sound is one way to audit this, but imo it's the more complicated way, even if we use privacy to further restrict the paths that can access the instance. We can take advantage of the particularities of this API to provide a tighter argument for soundness.

Reply:
Let's start by implementing `!Sync` for `WasmiInstance` to make sure this situation doesn't change under our feet.

Reply:
I was definitely taking this too lightly, I've added a doc test that `WasmInstance` is `!Sync`, and written a comment that describes our reasoning for why this is sound. Thanks for verifying this and helping me better elaborate on the safety here.